### PR TITLE
feat(store): improve create selector types

### PR DESCRIPTION
**Pull request by markwhitfeld**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently, the `createSelector` function relies on the types of the projector function's parameters to be specified explicitly.
This leaves room for mistakes to be made that would only be detected at runtime.

## What is the new behavior?

The projector function will now have its parameter types inferred from the selectors passed to the `createSelector` function.
This will allow for type errors to be detected at compile time, and for the types to even be removed from the projector function declaration in favor of the inferred types from the source selectors.

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

NOTE: Some code bases may experience compile errors after upgrading, but these errors would be identifying areas where the types are incorrect and need to be fixed.

### CODE REVIEW

File: packages/store/src/selectors/selector-metadata.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    selectorMetaData.containerClass = creationMetadata.containerClass;
+    selectorMetaData.selectorName = creationMetadata.selectorName || null;
+    getExplicitSelectorOptions =
+      creationMetadata.getSelectorOptions || getExplicitSelectorOptions;
+  }
+  const selectorMetaDataClone = { ...selectorMetaData };
+  selectorMetaData.getSelectorOptions = () =>
+    getLocalSelectorOptions(selectorMetaDataClone, getExplicitSelectorOptions());
+  return selectorMetaData;
+}
```
Comment (arturovt) [inline packages/store/src/selectors/selector-metadata.ts:37]: Can we have comments for this stuff? I think this is the darkest part of selectors implementation :smile:

File: packages/store/src/selectors/selector-utils.ts [hunk 2]
```diff
@@ -84,29 +40,24 @@ export function createSelector<T extends (...args: any[]) => any>(
...
+    const returnValue = originalFn.apply(containerClass, args);
+    if (returnValue instanceof Function) {
+      const innerMemoizedFn = memoize.apply(null, [returnValue]);
+      return innerMemoizedFn;
+    }
+    return returnValue;
+  } as T;
+  const memoizedFn = memoize(wrappedFn);
+  Object.setPrototypeOf(memoizedFn, originalFn);
+  return memoizedFn;
```
Comment (arturovt) [inline packages/store/src/selectors/selector-utils.ts:60]: Same here, can we have comments? Even me having troubles understanding it.

File: packages/store/src/selectors/selector-utils.ts [hunk 3]
```diff
@@ -84,29 +40,24 @@ export function createSelector<T extends (...args: any[]) => any>(
...
-    selectorMetaData.containerClass = creationMetadata.containerClass;
-    selectorMetaData.selectorName = creationMetadata.selectorName;
-    getExplicitSelectorOptions =
-      creationMetadata.getSelectorOptions || getExplicitSelectorOptions;
-  }
-  const selectorMetaDataClone = { ...selectorMetaData };
-  selectorMetaData.getSelectorOptions = () =>
-    getLocalSelectorOptions(selectorMetaDataClone, getExplicitSelectorOptions());
-  return selectorMetaData;
+  const containerClass = creationMetadata && creationMetadata.containerClass;
```
Comment (arturovt) [inline packages/store/src/selectors/selector-utils.ts:49]: We can use new features:
```suggestion
  const containerClass = creationMetadata?.containerClass;
```

**Review by arturovt — COMMENTED** [review]
I also updated Prettier rules on master so you can it sync it up and run the Prettier through your changed files. There should be less formatting changes.

File: packages/store/src/selectors/selector-utils.ts (hunk 2 above)
Comment (markwhitfeld) [reply packages/store/src/selectors/selector-utils.ts:60]: I can do this in a future PR. This is old code that has just been moved, and I wouldn't want it to block this PR.

**Review by arturovt — APPROVED** [review]

**Pull request merged**